lib/saluki-app/src/memory.rs

```diff
@@ -255,3 +270,68 @@ pub async fn initialize_allocator_telemetry() -> Result<(), GenericError> {
 
     Ok(())
 }
+
+struct CgroupMemoryParser {}
+
+impl CgroupMemoryParser {
+    fn parse(self) -> Option<ByteSize> {
+        if let Ok(mut file) = fs::File::open("/proc/self/cgroup") {
+            let mut contents = String::new();
+            if file.read_to_string(&mut contents).is_ok() {
+                let parts: Vec<&str> = contents.trim().split("\n").collect();
+                // CgroupV2 has unified controllers.
+                if parts.len() == 1 {
+                    return self.parse_controller_v2(parts[0]);
+                }
+                for line in contents.lines() {
+                    if line.contains(":memory:") {
+                        return self.parse_controller_v1(line);
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    /// Parse memory limit when CgroupV1 is in use.
+    ///
+    /// Defaults to None if any error is encountered.
+    fn parse_controller_v1(self, controller: &str) -> Option<ByteSize> {
+        if let Some(path) = controller.split(":").nth(2) {
+            let memory_path = format!("/sys/fs/cgroup/memory{}/memory.limit_in_bytes", path);
+            let memory_file_maybe = fs::File::open(memory_path).ok();
+            if let Some(mut memory_file) = memory_file_maybe {
+                let mut s = String::new();
+                if memory_file.read_to_string(&mut s).is_ok() {
+                    return self.convert_to_bytesize(&s);
+                }
+            }
+        }
+        None
```

Comment:
In Rust, you can also use the `?` operator for `Option<T>` return values in addition to `Result<T, E>`, which would let you get rid of almost all of the nesting here.

Taking advantage of that, plus some other simplifications, we could end up with something like this:

```rust
fn parse_controller_v1(self, controller: &str) -> Option<ByteSize> {
    let path = controller.split(":").nth(2)?;

    let memory_path = format!("/sys/fs/cgroup/memory{}/memory.limit_in_bytes", path);
    let raw_memory_limit = fs::read_to_string(memory_path).ok()?;
    self.convert_to_bytesize(&raw_memory_limit)
}
```